File: 0964_least_operators_to_express_number/solution.go
// LeetCode 0964 - Least Operators to Express Number
// https://leetcode.com/problems/least-operators-to-express-number/

func leastOpsExpressTarget(x int, target int) int {
	memo := map[int]int{}
	var dfs func(int) int
	dfs = func(t int) int {
		if v, ok := memo[t]; ok {
			return v
		}
		if x > t {
			a := 2*t - 1
			b := 2 * (x - t)
			if a < b {
				memo[t] = a
			} else {
				memo[t] = b
			}
			return memo[t]
		}
		if x == t {
			memo[t] = 0
			return 0
		}
		prod := x
		n := 0
		for prod < t {
			prod *= x
			n++
		}
		if prod == t {
			memo[t] = n
			return n
		}
		ans := dfs(t-prod/x) + n
		if prod < 2*t {
			cand := dfs(prod-t) + n + 1
			if cand < ans {
				ans = cand
			}
		}
		memo[t] = ans
		return ans
	}
	return dfs(target)
}

File: 0964_least_operators_to_express_number/solution.php
<?php
// LeetCode 0964 - Least Operators to Express Number
// https://leetcode.com/problems/least-operators-to-express-number/

class Solution {
    function leastOpsExpressTarget($x, $target) {
        $memo = [];
        $dfs = function ($t) use (&$dfs, &$memo, $x) {
            if (array_key_exists($t, $memo)) return $memo[$t];
            if ($x > $t) {
                $ans = min(2 * $t - 1, 2 * ($x - $t));
                $memo[$t] = $ans;
                return $ans;
            }
            if ($x === $t) {
                $memo[$t] = 0;
                return 0;
            }
            $prod = $x;
            $n = 0;
            while ($prod < $t) {
                $prod *= $x;
                $n++;
            }
            if ($prod === $t) {
                $memo[$t] = $n;
                return $n;
            }
            $ans = $dfs($t - intdiv($prod, $x)) + $n;
            if ($prod < 2 * $t) $ans = min($ans, $dfs($prod - $t) + $n + 1);
            $memo[$t] = $ans;
            return $ans;
        };
        return $dfs($target);
    }
}
